torchao/experimental/quant_api.py

```diff
@@ -463,3 +463,106 @@ def quantize(self, model: nn.Module) -> nn.Module:
             },
         )
         return model
+
+
+class IntxWeightOnlyQuantizedLinear(nn.Module):
+    def __init__(
+        self,
+        pack_weight_op,
+        linear_op,
+    ):
+        super().__init__()
+        self._pack_weights_op = pack_weight_op
+        self._linear_op = linear_op
+
+    def quantize_and_pack_weights(self, weights, nbit, group_size):
+        self.nbit = nbit
+        self.group_size = group_size
+
+        weight_qvals, weight_scales, weight_zeros = _quantize(
+            weights, self.group_size, self.nbit, True
+        )
+        weight_qvals = (weight_qvals + (1 << (nbit - 1))).to(torch.uint8)
+
+        self.weight_scales_and_zeros = torch.stack(
+            (weight_scales.t(), weight_zeros.t()), dim=2
+        )
+
+        self.packed_weights = self._pack_weights_op(weight_qvals.cpu()).to(device="mps")
+
+    def forward(self, x):
+        assert x.dim() == 2
+        return self._linear_op(
+            x, self.packed_weights, self.group_size, self.weight_scales_and_zeros
+        )
+
+
+def _replace_linear_with_quantized_linear_mps(module: nn.Module, kwargs={}):
+    group_size = kwargs["group_size"]
+    nbit = kwargs["nbit"]
+
+    assert not isinstance(module, nn.Linear)
+    assert nbit >= 1 and nbit <= 7
+
+    for name, child in module.named_children():
+        if not isinstance(child, nn.Linear):
+            _replace_linear_with_quantized_linear_mps(child, kwargs)
+        else:
+            assert child.bias is None
+            qlinear = IntxWeightOnlyQuantizedLinear(
+                pack_weight_op=getattr(torch.ops.torchao, f"_pack_weight_{nbit}bit"),
+                linear_op=getattr(
+                    torch.ops.torchao, f"_linear_fp_act_{nbit}bit_weight"
+                ),
+            )
+            setattr(module, name, qlinear)
+            getattr(module, name).quantize_and_pack_weights(
+                child.weight, nbit, group_size
+            )
+
+
+class IntxWeightOnlyLinearQuantizer:
```

Comment:
can you follow https://github.com/pytorch/ao/pull/995 instead

Reply:
We will converge towards adopting tensor subclass based API. In the meantime, we will use module swap to unblock us on e2e validation

Reply:
@manuelcandales let's file a task for this though

Reply:
done!